search.go

```diff
@@ -80,179 +70,46 @@ func newSearchFilterFromParams(r *http.Request) (searchFilter, error) {
-	span, ctx := apm.StartSpan(ctx, "Filter", "app")
-	defer span.End()
-
-	packagesList := map[string]map[string]util.Package{}
-
-	for _, p := range packages {
-		// Filter by category first as this could heavily reduce the number of packages
-		// It must happen before the version filtering as there only the newest version
-		// is exposed and there could be an older package with more versions.
-		if filter.Category != "" && !p.HasCategory(filter.Category) && !p.HasPolicyTemplateWithCategory(filter.Category) {
-			continue
-		}
-
-		// Skip internal packages
-		if p.Internal && !filter.Internal {
-			continue
-		}
-
-		// Skip experimental packages if flag is not specified
-		if p.Release == util.ReleaseExperimental && !filter.Experimental {
-			continue
-		}
-
-		if filter.KibanaVersion != nil {
-			if valid := p.HasKibanaVersion(filter.KibanaVersion); !valid {
-				continue
-			}
-		}
-
-		// If package Query is set, all versions of this package are returned
-		if filter.Package != "" && filter.Package != p.Name {
-			continue
-		}
-
-		addPackage := true
-		if !filter.AllVersions {
-			// Check if the version exists and if it should be added or not.
-			for name, versions := range packagesList {
-				if name != p.Name {
-					continue
-				}
-				for _, pp := range versions {
-
-					// If the package in the list is newer or equal, do nothing.
-					if pp.IsNewerOrEqual(p) {
-						addPackage = false
-						continue
-					}
-
-					// Otherwise delete and later add the new one.
-					delete(packagesList[pp.Name], pp.Version)
-				}
-			}
-		}
-
-		if addPackage {
-			if _, ok := packagesList[p.Name]; !ok {
-				packagesList[p.Name] = map[string]util.Package{}
-			}
-
-			if filter.Category != "" && !p.HasCategory(filter.Category) {
-				// It means that package's policy template has the category
-				p = filterPolicyTemplates(p, filter.Category)
-			}
-
-			if _, ok := packagesList[p.Name][p.Version]; !ok {
-				packagesList[p.Name][p.Version] = p
-			}
-		}
-	}
-	return packagesList
-}
-
-func filterPolicyTemplates(p util.Package, category string) util.Package {
-	var updatedPolicyTemplates []util.PolicyTemplate
-	var updatedBasePolicyTemplates []util.BasePolicyTemplate
-	for i, pt := range p.PolicyTemplates {
-		if util.StringsContains(pt.Categories, category) {
-			updatedPolicyTemplates = append(updatedPolicyTemplates, pt)
-			updatedBasePolicyTemplates = append(updatedBasePolicyTemplates, p.BasePackage.BasePolicyTemplates[i])
-		}
-	}
-	p.PolicyTemplates = updatedPolicyTemplates
-	p.BasePackage.BasePolicyTemplates = updatedBasePolicyTemplates
-	return p
-}
-
-func getPackageOutput(ctx context.Context, packagesList map[string]map[string]util.Package) ([]byte, error) {
+func getPackageOutput(ctx context.Context, packages util.Packages) ([]byte, error) {
 	span, ctx := apm.StartSpan(ctx, "GetPackageOutput", "app")
 	defer span.End()
 
-	separator := "@"
```

Comment:
I'm not sure if this logic wasn't placed here due to performance reasons. Maybe double-check it with @ruflin ?

Reply:
To what logic do you refer? To the whole search/filtering logic? I am moving it to a common place so all handlers have consistent behaviour, and we have a single place to look at to make performance optimizations if needed :slightly_smiling_face:

Reply:
I meant the sorting logic below.

Reply:
Well, before it was sorting the packages received as values in a map of maps, in that case sorting the packages as is done here would imply to copy the packages first to a list and then copying them again when sorting, or creating a list of pointers to the packages and sorting it. In any of these cases sorting only the keys and then creating the list of packages was probably better.

Now this method receives already a list of pointers to the packages, what can be sorted right away without the need of building the list of keys to sort.

Reply:
> Now this method receives already a list of pointers to the packages, what can be sorted right away without the need of building the list of keys to sort.

I was concerned that it operates on a list of relatively complex package structures. If this is not the thing, then feel free to resolve this comment.